src/cascadia/TerminalSettingsModel/ApplicationState.cpp

```diff
@@ -155,11 +193,18 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
     void ApplicationState::_write() const noexcept
     try
     {
-        Json::Value root{ Json::objectValue };
+        // re-read the state so that we can only update the properties that were changed.

```

Comment:
Is there a reason this change is necessary as part of this PR? Does it perhaps prevent any race condition from being too obvious?

Reply:
In retrospect, I added this after some of your comments about synchronization between different instances trying to write the file at the same time. Part of my concern at the time was that I didn't really understand how the state was getting refreshed (I missed the file change monitoring) and so I didn't want it to be the case that while we were waiting for the timer to go off another window saved the file and changes got lost. 

It can _probably_ be removed if you feel it is unnecessary. The only potential issue is that this automatic saving is the first, I think, usage of the application state where a write occurs without direct user input.

Reply:
Specifically, I wanted to address the case of one window trying to save, say, command line history, and if this window is trying to save the window layout they should not overwrite each other. I definitely think the solution can be improved upon. I'm less concerned about the window layout being overwritten on the automatic save (worst case the data is stale by a minute), as much as the window layout autosave clobbering some other user input.

Reply:
I may have further complicated this in #11222

Reply:
Things have changed a lot in the month since I wrote this part of the code. Again, I'm totally happy if things get refactored or outright deleted since I know I went from being the first user to one of 5 users of the application state.

Reply:
The application state wait is only in the time that the ApplicationState::SharedInstance is modified, not the whole time we are waiting for a save to happen.

E.g. the following is a potential race
1. User executes a commandline action from cmdpal
2. Simultaneously the monarch just collected all of its window layouts, and calls `ApplicationState::SharedInstance.PersistedWindowLayouts(windowLayouts)` which will trigger `_throttler` in `ApplicationState`.

Now, I don't know if the destructor for `ApplicationState` runs at the end of that statement (since the SharedInstance is out of scope), or if it is only called at application close. If the latter then there is a 1s wait from when the variable was set and when the write actually happens where another window could overwrite the persisted layout (or the commandline), or everything could be lost because the file gets reloaded every 100ms after a save.

Reply:
Went ahead and confirmed that the flush only happens when the application ends, not when the SharedInstance goes out of scope. 

I have the following proposal that I think will make this "safe enough" (doesn't handle merge conflicts, but at least doesn't clobber other changes most of the time):
1. Make the write throttler timeout something smaller, such as 100ms (or less) to match the file save refresh one.
2. When loading only load properties that are not marked dirty so that a reload doesn't overwrite unsaved changes from this file waiting on the throttler. 
3. When saving reload the file and only save the properties that we changed, then reset the dirty flags (current behavior).

This will still fail if two processes get to their `WriteUtf8FileAtomic`s at the same time.

The next step to this if the previous modifications are not enough without going to separate files per process entirely is
1. In _write check if `state.lock` exists, if it does call the _throttler again and return without saving (but don't clear the dirty flags)
2. After the WriteUtf8FileAtomic returns delete the `state.lock` file to allow another process to save their state.

Does this seem like a reasonable approach? I can probably throw together those changes quickly if it sounds good @lhecker .

Reply:
@Rosefield I like the idea of marking fields as dirty and only writing out changes.
Personally I'd leave the interval at 1s (that's still the current value right?).
But I'd use `LockFileEx` to block the background thread until an exclusive write access can be granted to `state.json`, _then_ read `state.json`, apply all dirty fields, and write it out to disk again.
That way we don't even need a `state.lock` file.

How do you like that idea? Please be honest though. 😄
If you'd like I can lend you a hand and code the file locking for you in another PR.

Reply:
BTW To solve this issue fundamentally I believe we'll have to use CRDTs, SQLite, or something like LevelDB/RocksDB in the future. SQLite would have the benefit of being shipped with Windows (as `winsqlite3.dll`).

Reply:
I didn't know `LockFileEx` existed, but I'm perfectly happy to utilize that instead. I've done minimal work with windows filesystems so all I knew is that I have different atomicity guarantees than linux filesystems where I'm used to seeing `.lock` files.

I'm perfectly happy to try to implement it myself, I'll let you know if I have any problems.

Reply:
Does this mean that ApplicationState should stop using the `{Read,Write}Utf8FileAtomic` methods?

Reply:
>  Personally I'd leave the interval at 1s (that's still the current value right?).

Yes, 1s is still the current value. Is there a case where having this value be 100ms causes issues? I doubt we'd actually end up with more than 1 write on the same process per 1s, but flushing sooner means that other processes get the updates quicker.

Reply:
> Does this mean that ApplicationState should stop using the `{Read,Write}Utf8FileAtomic` methods?

Yeah I think so.

> Is there a case where having this value be 100ms causes issues?

The reason I built it that way isn't just because it makes the usage simpler, but also because I wanted to be able to call `ApplicationState` functions as often as the display refresh rate without writing to disk too often. For instance imagine (this is just a hypothetical example) we'd store the window position while the window is dragged: If we set the interval to 100ms, we'd write to disk quite often, which I'd like to prevent.

Reply:
Ah I forgot to add: I believe we can use [Mailslots](https://docs.microsoft.com/en-us/windows/win32/ipc/mailslots) in the future to achieve near-instant refresh across processes. Alternatively we could simply move the `ApplicationState` singleton over to the monarch process. Since COM is agnostic over whether the target class is in the current process or not, it would just work ™️.

Reply:
Both sound like reasonable followups to this :)